### PR TITLE
cmake changes

### PR DESCRIPTION
**Pull request by pca006132**
I was looking at the changes required to better support including this project as a subdirectory in a cmake project, and went down the rabbit hole of changing stuff...

This PR is very radical, and I do not expect everything to be merged as-is. This is mostly for testing in the CI as well as gather some feedback from the users.

@starseeker @kintel will be nice if you can provide some feedback. I will comment on things that are a bit more interesting than moving things around.

### CODE REVIEW

File: bindings/python/CMakeLists.txt
Comment (pca006132) [reply bindings/python/CMakeLists.txt:48]: and I thought of moving this to `cmake/manifoldDeps.cmake`, but not sure if that is the best thing to do...

**Comment by elalish** [comment]
This looks like the last PR for v3.0 - agreed? I don't know enough CMake to review this one particularly, but it certainly looks to be going in a good direction. How much feedback do you want to wait for before we merge? And after all the commits, do you still feel this is a preliminary attempt?

**Comment by pca006132** [comment]
Changes here are mostly internal, except the target alias that is already covered by #1049. I also changed `MANIFOLD_CBIND` a bit to allow it to build as a static library, which rust users may be interested in. But again this can be moved to a separate PR, and that is not related to system library anyway, so missing a release is fine.

Also, this includes things from #1045, which is probably not that ready for a release. I am fine for merging this after v3.0, and get some feedback from the users, who have more experience working with cmake than I do.

File: CMakeLists.txt
Comment (kintel) [inline CMakeLists.txt:150]: Are there compiler flags now supported across all supported versions of (gcc and clang) compilers?

**Comment by kintel** [comment]
One think that I'm uncertain how to deal with:
1) Build and install manifold using the built-in clipper
2) Update the version of the built-in clipper
3) Build and reinstall manifold -> it will pick up the clipper from the previous installation instead of using the built-in one

This can be solved by never using builtins, but it feels wrong the way it's done today. Not sure how easy it is to work around this, either from within manifold or inside the target build environment.

**Comment by elalish** [comment]
Agreed about risk - let's hold this to after v3.0. I'll start working on the release.

**Comment by kintel** [comment]
Otoh, if this isn't addressed before the 3.0 release, people may have to resort to using Manifold from HEAD instead of a released version, which will just delay integrations and cause messy workarounds in client code.

**Comment by pca006132** [comment]
> One think that I'm uncertain how to deal with:
> 
>     1. Build and install manifold using the built-in clipper
> 
>     2. Update the version of the built-in clipper
> 
>     3. Build and reinstall manifold -> it will pick up the clipper from the previous installation instead of using the built-in one
> 
> 
> This can be solved by never using builtins, but it feels wrong the way it's done today. Not sure how easy it is to work around this, either from within manifold or inside the target build environment.

I see. One way is to always build clipper as static library when using the builtin one, so we don't have to install it on the system.

File: CMakeLists.txt
Comment (pca006132) [reply CMakeLists.txt:150]: I think the most important flag is `-Wno-unknown-warning-option`, it allows us to use some other warning options not supported by the compiler. We should check when gcc and clang supports this...

Comment (pca006132) [reply CMakeLists.txt:150]: This seems to be a clang only flag, which is supported at least 13 years ago (https://bugzilla.mozilla.org/show_bug.cgi?id=731316).

GCC will not complain about unknown warning options anyway:

> When an unrecognized warning option is requested (e.g., -Wunknown-warning), GCC emits a diagnostic stating that the option is not recognized. However, if the -Wno- form is used, the behavior is slightly different: no diagnostic is produced for -Wno-unknown-warning unless other diagnostics are being produced. This allows the use of new -Wno- options with old compilers, but if something goes wrong, the compiler warns that an unrecognized option is present. 

These flags should be fine.

**Comment by pca006132** [comment]
I can backport the dependency changes to another PR for v3.0.

The major concern I have is that tbb doesn't officially support building as a static library. Users concerned about this should make sure they are not using the built-in tbb. We are also doing this for python as well, although this is easier to fix for python if needed...

File: cmake/manifoldDeps.cmake
Comment (kintel) [inline cmake/manifoldDeps.cmake:38]: Note: When building as static libs, the libraries become bundled into the dynamic libmanifold library. This means that any attempt to locate the static library will fail, so things like `find_package(Clipper2 REQUIRED)` in `manifoldConfig.cmake` would also need to be removed for this build variant

Comment (kintel) [reply cmake/manifoldDeps.cmake:38]: ..and we need to remove those libraries from the public link interface of the manifold target (PUBLIC -> PRIVATE)

**Comment by kintel** [comment]
One idea after trying (and still failing) to use this PR: Write a test that builds and installs manifold into some local install folder, then build a simple example program linking against the installed manifold. That would test that manifoldConfig.cmake does the right thing. Similar tests could be written for pkg-config, as well as build variants. It might be a bit annoying to set this up the first time, but it could be a powerful tool to sleep better at night around release time.

**Comment by pca006132** [comment]
We do have cmake consumer test, the issue is that we are not testing different variants of this, and there are too many variants... But yeah we should probably do that.

Another approach is, maybe we can take the easier route to just not install manifold into the system if we are using built-in libraries? These are mainly for easy development, and there are so many ways installation can go wrong... Similarly we may set MANIFOLD_DOWNLOAD to off when we are not the top level project, and let the parent projecr handle all dependency issues.

**Comment by pca006132** [comment]
@kintel and by failing to use, what is your configuration?

**Comment by kintel** [comment]
For testing I was installing manifold from source into the system, then trying to build OpenSCAD. All on macOS: https://github.com/openscad/openscad/pull/5310

**Comment by kintel** [comment]
With the latest changes it works a lot better for me on macOS (building OpenSCAD on macOS for binary distribution).

**Comment by pca006132** [comment]
Finally, all green!!! @kintel let me know if you got any issues from your testing. I now added several more package consumers, either using the installed library or from a subdirectory.

@hzeller do you know if it is normal to have rpath flags in .pc file? I need this to make it work on macos, apparently the installation path is not in the default loader path...

**Comment by hzeller** [comment]
Mostly the package config files provide whatever flags are needed for use in cflags and libs to get things compiled. So if macos requires rpaths, then they should probably be there. I don't really know anything about macs, but needing rpaths sounds like a brute-foce way of providing dependencies (well, at least I never had to do that in *.pc files, so I can't really comment what a good approach is).
Maybe there is someone who would like to build a brew package and can figure out mac-related details ?

**Comment by pca006132** [comment]
> Mostly the package config files provide whatever flags are needed for use in cflags and libs to get things compiled. So if macos requires rpaths, then they should probably be there. I don't really know anything about macs, but needing rpaths sounds like a brute-foce way of providing dependencies (well, at least I never had to do that in *.pc files, so I can't really comment what a good approach is). Maybe there is someone who would like to build a brew package and can figure out mac-related details ?

Looking at some of the *.pc files I have, they don't seem to have this rpath setting included. It seems that cmake handles rpath automatically, but you may need to handle this by yourself if you call the compiler directly?

Anyway, this is left as an exercise to users using pkg-config on macos I guess...

**Comment by kintel** [comment]
I cannot speak for other platforms, but the nice thing to do when building for macOS is to set `LC_ID_DYLIB` to `@rpath/libmanifold.2.dylib`, which seems to be the case for manifold (I assume that's what modern CMake does).
This will make the library resolve to whatever rpath the application sets up, which in the case of macOS is typically either a package manager like Homebrew, or bundled libraries (which OpenSCAD does).

File: bindings/wasm/bindings.js
Comment (elalish) [inline bindings/wasm/bindings.js:22]: Is this including changes from the emscripten TBB PR? Should they be separate?

File: cmake/manifoldDeps.cmake
Comment (elalish) [inline cmake/manifoldDeps.cmake:110]: Does this mean we're now auto-fetching dependencies again? How does this relate to our CMake flag for fetching dependencies?

File: src/parallel.h
Comment (elalish) [inline src/parallel.h:1]: I thought you had put this in include so OpenSCAD could use our parallel functions. Did something change?

**Review by elalish — CHANGES_REQUESTED** [review]
So, it looks like we want to merge this for v3.0 - but I have a few questions first:

File: cmake/manifoldDeps.cmake
Comment (pca006132) [reply cmake/manifoldDeps.cmake:110]: No, the `logmissingdep` function check the cmake flag and will error when we disabled download.

File: src/parallel.h
Comment (pca006132) [reply src/parallel.h:1]: Yes this was the plan, but later I found that I can just use tbb functions directly there. Having this as a public header will require us to make tbb a public dependency, which is problematic when we build the fetched dependency as a static library...

File: bindings/wasm/bindings.js
Comment (pca006132) [reply bindings/wasm/bindings.js:22]: Yes I can remove them.

**Comment by pca006132** [comment]
> I cannot speak for other platforms, but the nice thing to do when building for macOS is to set `LC_ID_DYLIB` to `@rpath/libmanifold.2.dylib`, which seems to be the case for manifold (I assume that's what modern CMake does). This will make the library resolve to whatever rpath the application sets up, which in the case of macOS is typically either a package manager like Homebrew, or bundled libraries (which OpenSCAD does).

Would be great if someone with mac can check if our build already does that. @elalish @kintel

File: src/parallel.h
Comment (elalish) [reply src/parallel.h:1]: Okay, sounds good to me - the less in our public includes the better, as far as I'm concerned.

**Comment by elalish** [comment]
I have a mac, but I'm not sure what to check. Also, I'm going to bed.

**Comment by pca006132** [comment]
@hzeller I moved the `version.h.in` into `cmake` directory (not sure if it is the right place for this kind of things), as well as modified the build interface include path of `manifold` target to include `${PROJECT_BINARY_DIR}/include`, so the generated `version.h` can be referenced regardless of whether the library is used from a subdirectory or as an installed library.

**Comment by pca006132** [comment]
@elalish I guess we can wait for kintel to test that :)

**Comment by hzeller** [comment]
> @hzeller I moved the `version.h.in` into `cmake` directory (not sure if it is the right place for this kind of things), as well as modified the build interface include path of `manifold` target to include `${PROJECT_BINARY_DIR}/include`, so the generated `version.h` can be referenced regardless of whether the library is used from a subdirectory or as an installed library.

SGTM, thanks!

**Review by elalish — APPROVED** [review]
LGTM, though I'm no CMake expert. I'll go update our required tests so merging isn't blocked.

File: CMakeLists.txt
Comment (pca006132) [inline CMakeLists.txt:78]: Users can choose not to use the system version of the dependencies. I saw this in the openscad wishlist, so I added it.

**Comment by pca006132** [comment]
Exported library and binary files are now placed under `lib/Release` and `bin/Release` respectively under msvc, i.e. `manifold.lib`, `manifoldc.lib`, `manifold_test.exe` and various executables under `extras`.

Configuration info is now logging the full set of compile and link options for manifold for better debugging...

**Comment by kintel** [comment]
> 
> Would be great if someone with mac can check if our build already does that. @elalish @kintel

Look good from here,
I have TBB installed on the system but not Clipper:

```
$  cmake -DCMAKE_INSTALL_PREFIX=install -DCMAKE_OSX_DEPLOYMENT_TARGET=10.15 -DCMAKE_OSX_ARCHITECTURES="x86_64 arm64" -DMANIFOLD_CBIND=OFF -DMANIFOLD_TEST=OFF -DMANIFOLD_PAR=ON ..
$  make install
```

```
otool -L install/lib/libmanifold.dylib
install/lib/libmanifold.dylib (architecture x86_64):
	@rpath/libmanifold.2.dylib (compatibility version 2.0.0, current version 2.5.1)
	@rpath/libtbb.12.dylib (compatibility version 12.0.0, current version 12.12.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
install/lib/libmanifold.dylib (architecture arm64):
	@rpath/libmanifold.2.dylib (compatibility version 2.0.0, current version 2.5.1)
	@rpath/libtbb.12.dylib (compatibility version 12.0.0, current version 12.12.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1336.61.1)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 1600.157.0)
```

**Comment by pca006132** [comment]
I think this should be ready to merge?

File: cmake/version.h.in
Comment (kintel) [inline cmake/version.h.in:19]: Typo: PATH -> PATCH

**Comment by kintel** [comment]
Re. compile time version. It could be nice with a single integer version number in addition. I can find examples from other libs if needed for inspiration

File: cmake/version.h.in
Comment (hzeller) [reply cmake/version.h.in:19]: Ups, I take full responsibility for that typo...

**Comment by hzeller** [comment]
I think this should be merged soon, as it is getting big and allows folks to easier play with it, and final polish-ups can be done in a follow-up.

**Comment by kintel** [comment]
> Re. compile time version. It could be nice with a single integer version number in addition. I can find examples from other libs if needed for inspiration

Examples:
```CGAL_VERSION_NR < CGAL_VERSION_NUMBER(5, 4, 0)

**Comment by elalish** [comment]
Okay, I fixed the typo, so I'm going to go ahead and merge this. I'm not quite sure how you want a single version number int exposed, so I'll wait for a PR from @kintel or @hzeller for that. When it sounds like things have settled, I'll cut v3.0. Please let me know - I'd like to avoid an immediate 3.0.1 if possible.

**Comment by hzeller** [comment]
I have put the comparable version number suggestion in https://github.com/elalish/manifold/pull/1054

File: CMakeLists.txt
Comment (starseeker) [inline CMakeLists.txt:267]: Slick!  That's a new one for me.

File: bindings/python/CMakeLists.txt
Comment (starseeker) [inline bindings/python/CMakeLists.txt:46]: Ah, good catch.  Using CMAKE_SOURCE_DIR is a habit I have left from the old days, looks like PROJECT_SOURCE_DIR is probably what one would want most of the time.

File: src/parallel.h
Comment (starseeker) [reply src/parallel.h:1]: I wasn't plugged in this weekend, so it's too late for 3.0, but  IIRC iters.h was only public in the headers for parallel.h so as long as that's still true we can make that header private too for the next round.

**Pull request merged**
